HighlightUtilities/functions.php

```diff
@@ -196,56 +196,59 @@ function getLanguageDefinitionPath($name)
  * @since 9.15.6.1
  *
  * @param string $html An HTML string generated by `Highlighter::highlight()`
  *
- * @throws \RuntimeException         when the DOM extension is not available
- * @throws \UnexpectedValueException when the given HTML could not be parsed
- *
  * @return string[]|false An array of lines of code as strings. False if an error occurred in splitting up by lines
  */
 function splitCodeIntoArray($html)
 {
-    if (!extension_loaded("dom")) {
-        throw new \RuntimeException("The DOM extension is not loaded but is required.");
-    }
-
     if (trim($html) === "") {
         return array();
     }
 
-    $dom = new \DOMDocument();
+    $queuedPrefix = '';
+    $regexWorkspace = array();
+    $rawLines = preg_split('/\R/u', $html);
 
-    // https://stackoverflow.com/a/8218649
-    if (!$dom->loadHTML(mb_convert_encoding($html, "HTML-ENTITIES", "UTF-8"))) {
-        throw new \UnexpectedValueException("The given HTML could not be parsed correctly.");
+    if ($rawLines === false) {
+        return false;
     }
 
-    $xpath = new \DOMXPath($dom);
-    $spans = $xpath->query("//span[contains(text(), '\n') or contains(text(), '\r\n')]");
+    foreach ($rawLines as &$rawLine) {
+        // If the previous line has been marked as "open", then we'll have something
+        // in our queue
+        if ($queuedPrefix !== '') {
+            $rawLine = $queuedPrefix . $rawLine;
+            $queuedPrefix = '';
+        }
 
-    /** @var \DOMElement $span */
-    foreach ($spans as $span) {
-        $closingTags = '';
-        $openingTags = '';
-        $curr = $span;
+        // Find how many opening `<span>` tags exist on this line
+        preg_match_all('/<span[^<>]+>/u', $rawLine, $regexWorkspace);
+        $openingTags = count($regexWorkspace[0]);
 
-        while ($curr->tagName === 'span') {
-            $closingTags .= '</span>';
-            $openingTags = sprintf('<span class="%s">%s', $curr->getAttribute("class"), $openingTags);
+        // Find all of the closing `</span>` tags that exist on this line
+        preg_match_all('/<\/span>/u', $rawLine, $regexWorkspace);
+        $closingTags = count($regexWorkspace[0]);
 
-            $curr = $curr->parentNode;
+        // If the number of opening tags matches the number of closing tags, then
+        // we don't have any new tags that span multiple lines
+        if ($openingTags === $closingTags) {
+            continue;
         }
 
-        $renderedSpan = $dom->saveHTML($span);
-        $finished = preg_replace(
-            '/\R/u',
-            $closingTags . PHP_EOL . $openingTags,
-            $renderedSpan
-        );
-        $html = str_replace($renderedSpan, $finished, $html);
+        // Find all of the complete `<span>` tags and remove them from a working
+        // copy of the line. Then we'll be left with just opening tags.
+        $workingLine = preg_replace('/<span[^<]+>[^<>]+<\/span>/u', '', $rawLine);
```

Comment:
Shouldn't this be able to match an empty `span`?

```suggestion
        $workingLine = preg_replace('/<span[^<]*+>[^<]*+<\/span>/u', '', $rawLine);
```

Reply:
I edited your suggestion to invert the `[^<]` to `[^>]` so I could batch commit via GH; was an original typo on my part